internal/addonsdk/sdk.go

```diff
@@ -0,0 +1,215 @@
+package addonsdk
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/go-logr/logr"
+	addonsv1alpha1 "github.com/openshift/addon-operator/apis/addons/v1alpha1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+var (
+	statusReporterSingleton      *StatusReporter
+	statusReporterSingletonMutex = &sync.Mutex{}
+)
+
+// a nice heartbeat-reporter implemented by the MT-SRE which our tenants can use
+// if they don't like it, they can implement their own heartbeat reporter by creating a type which implements the `addonsdk.statusReporterClient` interface
+type StatusReporter struct {
+	// object provided by the client/tenants which implements the addonsdk.client interface
+	addonInstanceInteractor client
+	addonName               string
+	addonTargetNamespace    string
+
+	// the latest conditions which the heartbeat reporter would be reporting periodically
+	latestConditions []metav1.Condition
+
+	// to control the rate at which the heartbeat reporter would run
+	interval time.Duration
+	ticker   *time.Ticker
+
+	// for effectively communicating the stop and update signals
+	stopperCh chan bool
+	updateCh  chan updateOptions
+
+	// for concurrency-safely executing one instance of heartbeat reporter loop
+	executeOnce sync.Once
+
+	//for tracking if the heartbeat reporter is running or done running
+	doneCh chan bool
+
+	log logr.Logger
+}
+
+// ensure that the `StatusReporter` implements the `addonsdk.statusReporterClient` interface
+var _ statusReporterClient = (*StatusReporter)(nil)
+
+// InitializeStatusReporterSingleton sets up a singleton of the type `StatusReporter` (only if it doesn't exist yet) and returns it to the caller.
+func InitializeStatusReporterSingleton(addonInstanceInteractor client, addonName string, addonTargetNamespace string, logger logr.Logger) *StatusReporter {
+	if statusReporterSingleton == nil {
+		statusReporterSingletonMutex.Lock()
+		defer statusReporterSingletonMutex.Unlock()
+		if statusReporterSingleton == nil {
+			statusReporterSingleton = &StatusReporter{
+				addonInstanceInteractor: addonInstanceInteractor,
+				addonName:               addonName,
+				addonTargetNamespace:    addonTargetNamespace,
+				latestConditions: []metav1.Condition{
+					{
+						Type:    "addons.managed.openshift.io/Healthy",
+						Status:  "False",
+						Reason:  "NoHeartbeatReported",
+						Message: fmt.Sprintf("Addon '%s' hasn't reported any heartbeat yet", addonName),
+					},
+				},
+				stopperCh: make(chan bool),
+				updateCh:  make(chan updateOptions),
+				doneCh:    make(chan bool),
+				log:       logger,
+			}
+			// because the heartbeat reporter still hasn't been started
+			defer close(statusReporterSingleton.doneCh)
```

Comment:
Yep, because calling this method function just creates/gets the Singleton object, not start. 
Hence, closing this channel right away in edge cases where a client creates/gets this singleton object and straight away, without starting it, calls `.Stop()` or `.SetConditions` or `.ReportAddonInstanceSpecChange()` methods on it.

For starting the status reporter, the `StatusReporter.Start(ctx) error` method has to be explicitly called.

Reply:
This again is a data race, if you really want to implement this in this fashion, you need an extra mutex to lock access to the `doneCh` property or the behavior of concurrent access to it will be undefined.

Reply:
But this entire sub-function is locked safely inside Once.Do, thereby not allowing any concurrent access to it and this `defer close()` defers to this sub-function and not the `SetupStatusReporter` i.e. parent function which is still exposed to concurrent access.